File: Java/string_utils/StringUtilsTest.java
package string_utils;

import java.util.*;

/**
 * StringUtils 单元测试
 * 
 * 运行方式：
 * cd Java/string_utils
 * javac *.java && java string_utils.StringUtilsTest
 */
public class StringUtilsTest {
    
    private static int passed = 0;
    private static int failed = 0;
    
    public static void main(String[] args) {
        System.out.println("=== StringUtils Test Suite ===\n");
        
        testCaseConversion();
        testTrimAndPad();
        testTruncate();
        testInterpolation();
        testEscaping();
        testExtraction();
        testCharStats();
        testStringOperations();
        testBase64();
        testSimilarity();
        testUtilities();
        
        System.out.println("\n=== Test Summary ===");
        System.out.println("Passed: " + passed);
        System.out.println("Failed: " + failed);
        System.out.println("Total:  " + (passed + failed));
        
        if (failed > 0) {
            System.exit(1);
        }
    }
    
    private static void assertTrue(String message, boolean condition) {
        if (condition) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAILED: " + message);
        }
    }
    
    private static void assertEquals(String message, String expected, String actual) {
        if (Objects.equals(expected, actual)) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAILED: " + message);
            System.out.println("    Expected: " + expected);
            System.out.println("    Actual:   " + actual);
        }
    }
    
    private static void assertEquals(String message, int expected, int actual) {
        if (expected == actual) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAILED: " + message);
            System.out.println("    Expected: " + expected);
            System.out.println("    Actual:   " + actual);
        }
    }
    
    private static void assertEquals(String message, double expected, double actual, double delta) {
        if (Math.abs(expected - actual) <= delta) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAILED: " + message);
            System.out.println("    Expected: " + expected);
            System.out.println("    Actual:   " + actual);
        }
    }
    
    private static void testCaseConversion() {
        System.out.println("Testing Case Conversion...");
        
        // camelCase
        assertEquals("toCamelCase('hello_world')", "helloWorld", StringUtils.toCamelCase("hello_world"));
        assertEquals("toCamelCase('Hello-World-Test')", "helloWorldTest", StringUtils.toCamelCase("Hello-World-Test"));
        assertEquals("toCamelCase('alreadyCamelCase')", "alreadyCamelCase", StringUtils.toCamelCase("alreadyCamelCase"));
        
        // PascalCase
        assertEquals("toPascalCase('hello_world')", "HelloWorld", StringUtils.toPascalCase("hello_world"));
        assertEquals("toPascalCase('hello-world')", "HelloWorld", StringUtils.toPascalCase("hello-world"));
        
        // snake_case
        assertEquals("toSnakeCase('helloWorld')", "hello_world", StringUtils.toSnakeCase("helloWorld"));
        assertEquals("toSnakeCase('HelloWorldTest')", "hello_world_test", StringUtils.toSnakeCase("HelloWorldTest"));
        
        // kebab-case
        assertEquals("toKebabCase('helloWorld')", "hello-world", StringUtils.toKebabCase("helloWorld"));
        assertEquals("toKebabCase('Hello World')", "hello-world", StringUtils.toKebabCase("Hello World"));
        
        // CONSTANT_CASE
        assertEquals("toConstantCase('helloWorld')", "HELLO_WORLD", StringUtils.toConstantCase("helloWorld"));
        
        // dot.case
        assertEquals("toDotCase('helloWorld')", "hello.world", StringUtils.toDotCase("helloWorld"));
        
        // path/case
        assertEquals("toPathCase('helloWorld')", "hello/world", StringUtils.toPathCase("helloWorld"));
        
        // space case
        assertEquals("toSpaceCase('helloWorld')", "hello world", StringUtils.toSpaceCase("helloWorld"));
        
        // Sentence case
        assertEquals("toSentenceCase('hello WORLD')", "Hello world", StringUtils.toSentenceCase("hello WORLD"));
        
        // Edge cases
        assertEquals("toCamelCase(null)", null, StringUtils.toCamelCase(null));
        assertEquals("toCamelCase('')", "", StringUtils.toCamelCase(""));
        assertEquals("toCamelCase('a')", "a", StringUtils.toCamelCase("a"));
        
        System.out.println("  Case Conversion: PASSED\n");
    }
    
    private static void testTrimAndPad() {
        System.out.println("Testing Trim and Pad...");
        
        // trim
        assertEquals("trim('  hello  ')", "hello", StringUtils.trim("  hello  "));
        assertEquals("trim(null)", null, StringUtils.trim(null));
        
        // trimLeft
        assertEquals("trimLeft('  hello  ')", "hello  ", StringUtils.trimLeft("  hello  "));
        
        // trimRight
        assertEquals("trimRight('  hello  ')", "  hello", StringUtils.trimRight("  hello  "));
        
        // padLeft
        assertEquals("padLeft('123', 5, '0')", "00123", StringUtils.padLeft("123", 5, '0'));
        assertEquals("padLeft('12345', 3, '0')", "12345", StringUtils.padLeft("12345", 3, '0'));
        
        // padRight
        assertEquals("padRight('123', 5, ' ')", "123  ", StringUtils.padRight("123", 5, ' '));
        
        // zeroPad
        assertEquals("zeroPad('42', 4)", "0042", StringUtils.zeroPad("42", 4));
        
        // center
        assertEquals("center('hi', 5, '*')", "*hi**", StringUtils.center("hi", 5, '*'));
        assertEquals("center('hi', 4, '*')", "*hi*", StringUtils.center("hi", 4, '*'));
        
        System.out.println("  Trim and Pad: PASSED\n");
    }
    
    private static void testTruncate() {
        System.out.println("Testing Truncate...");
        
        // truncate with suffix
        assertEquals("truncate('Hello World', 8, '...')", "Hello...", StringUtils.truncate("Hello World", 8, "..."));
        assertEquals("truncate('Hi', 10, '...')", "Hi", StringUtils.truncate("Hi", 10, "..."));
        
        // truncate default
        assertEquals("truncate('Hello World', 8)", "Hello...", StringUtils.truncate("Hello World", 8));
        
        // truncateWords
        assertEquals("truncateWords('Hello World Test', 2, '...')", "Hello World...", StringUtils.truncateWords("Hello World Test", 2, "..."));
        assertEquals("truncateWords('Hello World', 5, '...')", "Hello World", StringUtils.truncateWords("Hello World", 5, "..."));
        
        System.out.println("  Truncate: PASSED\n");
    }
    
    private static void testInterpolation() {
        System.out.println("Testing Interpolation...");
        
        // interpolate
        Map<String, String> vars = new HashMap<>();
        vars.put("name", "World");
        vars.put("greeting", "Hello");
        assertEquals("interpolate", "Hello World!", StringUtils.interpolate("${greeting} ${name}!", vars));
        assertEquals("interpolate missing", "${missing}", StringUtils.interpolate("${missing}", vars));
        
        // format
        assertEquals("format", "Hello World", StringUtils.format("{0} {1}", "Hello", "World"));
        assertEquals("format numbers", "Value: 42", StringUtils.format("Value: {0}", 42));
        
        System.out.println("  Interpolation: PASSED\n");
    }
    
    private static void testEscaping() {
        System.out.println("Testing Escaping...");
        
        // escapeHtml
        assertEquals("escapeHtml", "&lt;script&gt;", StringUtils.escapeHtml("<script>"));
        assertEquals("escapeHtml quotes", "&quot;test&quot;", StringUtils.escapeHtml("\"test\""));
        
        // unescapeHtml
        assertEquals("unescapeHtml", "<script>", StringUtils.unescapeHtml("&lt;script&gt;"));
        
        // escapeXml
        assertEquals("escapeXml", "&apos;test&apos;", StringUtils.escapeXml("'test'"));
        
        // escapeJson
        assertEquals("escapeJson newline", "\\n", StringUtils.escapeJson("\n"));
        assertEquals("escapeJson quote", "\\\"test\\\"", StringUtils.escapeJson("\"test\""));
        
        // escapeSql
        assertEquals("escapeSql", "O''Brien", StringUtils.escapeSql("O'Brien"));
        
        // escapeRegex
        assertEquals("escapeRegex", "\\Q.test\\E", StringUtils.escapeRegex(".test"));
        
        // urlEncode/Decode
        assertEquals("urlEncode", "Hello+World", StringUtils.urlEncode("Hello World"));
        assertEquals("urlDecode", "Hello World", StringUtils.urlDecode("Hello+World"));
        
        System.out.println("  Escaping: PASSED\n");
    }
    
    private static void testExtraction() {
        System.out.println("Testing Extraction...");
        
        String text = "Contact us at test@example.com or support@domain.org. Visit https://example.com or call +1234567890. #trending @user";
        
        // extractEmails
        List<String> emails = StringUtils.extractEmails(text);
        assertEquals("extractEmails count", 2, emails.size());
        assertTrue("extractEmails contains", emails.contains("test@example.com"));
        
        // extractUrls
        List<String> urls = StringUtils.extractUrls(text);
        assertEquals("extractUrls count", 1, urls.size());
        assertTrue("extractUrls contains", urls.contains("https://example.com"));
        
        // extractPhoneNumbers
        List<String> phones = StringUtils.extractPhoneNumbers(text);
        assertEquals("extractPhoneNumbers count", 1, phones.size());
        
        // extractHashtags
        List<String> hashtags = StringUtils.extractHashtags(text);
        assertEquals("extractHashtags count", 1, hashtags.size());
        assertEquals("extractHashtags value", "trending", hashtags.get(0));
        
        // extractMentions
        List<String> mentions = StringUtils.extractMentions(text);
        assertEquals("extractMentions count", 1, mentions.size());
        assertEquals("extractMentions value", "user", mentions.get(0));
        
        // extractNumbers
        List<String> numbers = StringUtils.extractNumbers("Price: $100, discount 20%, final 80");
        assertEquals("extractNumbers count", 3, numbers.size());
        
        System.out.println("  Extraction: PASSED\n");
    }
    
    private static void testCharStats() {
        System.out.println("Testing Character Stats...");
        
        String str = "Hello 123!";
        
        assertEquals("countLetters", 5, StringUtils.countLetters(str));
        assertEquals("countDigits", 3, StringUtils.countDigits(str));
        assertEquals("countWhitespace", 1, StringUtils.countWhitespace(str));
        assertEquals("countSpecialChars", 1, StringUtils.countSpecialChars(str));
        
        Map<String, Integer> stats = StringUtils.getCharStats(str);
        assertEquals("stats total", 10, stats.get("total").intValue());
        assertEquals("stats letters", 5, stats.get("letters").intValue());
        
        System.out.println("  Character Stats: PASSED\n");
    }
    
    private static void testStringOperations() {
        System.out.println("Testing String Operations...");
        
        // reverse
        assertEquals("reverse", "olleH", StringUtils.reverse("Hello"));
        
        // replace
        assertEquals("replace", "Hi Hi Hi", StringUtils.replace("Hello Hello Hello", "Hello", "Hi", -1));
        assertEquals("replace limited", "Hi Hi Hello", StringUtils.replace("Hello Hello Hello", "Hello", "Hi", 2));
        
        // insert
        assertEquals("insert", "HelXlo", StringUtils.insert("Hello", "X", 3));
        
        // delete
        assertEquals("delete", "Helo", StringUtils.delete("Hello", 2, 3));
        
        // repeat
        assertEquals("repeat", "abcabcabc", StringUtils.repeat("abc", 3));
        
        // capitalize
        assertEquals("capitalize", "Hello", StringUtils.capitalize("hello"));
        
        // uncapitalize
        assertEquals("uncapitalize", "hello", StringUtils.uncapitalize("Hello"));
        
        // swapCase
        assertEquals("swapCase", "hELLO", StringUtils.swapCase("Hello"));
        
        // isPalindrome
        assertTrue("isPalindrome", StringUtils.isPalindrome("A man a plan a canal Panama"));
        assertTrue("isPalindrome number", StringUtils.isPalindrome("12321"));
        assertFalse("isPalindrome false", StringUtils.isPalindrome("hello"));
        
        // join
        assertEquals("join array", "a,b,c", StringUtils.join(",", "a", "b", "c"));
        assertEquals("join collection", "x y z", StringUtils.join(" ", Arrays.asList("x", "y", "z")));
        
        // split
        List<String> parts = StringUtils.split("a,b,c", ",");
        assertEquals("split count", 3, parts.size());
        
        System.out.println("  String Operations: PASSED\n");
    }
    
    private static void testBase64() {
        System.out.println("Testing Base64...");
        
        String original = "Hello, World! 你好世界";
        
        // base64Encode/Decode
        String encoded = StringUtils.base64Encode(original);
        String decoded = StringUtils.base64Decode(encoded);
        assertEquals("base64 roundtrip", original, decoded);
        
        // base64UrlEncode/Decode
        String urlEncoded = StringUtils.base64UrlEncode(original);
        String urlDecoded = StringUtils.base64UrlDecode(urlEncoded);
        assertEquals("base64Url roundtrip", original, urlDecoded);
        
        // null handling
        assertEquals("base64Encode null", null, StringUtils.base64Encode(null));
        assertEquals("base64Decode null", null, StringUtils.base64Decode(null));
        
        System.out.println("  Base64: PASSED\n");
    }
    
    private static void testSimilarity() {
        System.out.println("Testing Similarity...");
        
        // levenshteinDistance
        assertEquals("levenshtein identical", 0, StringUtils.levenshteinDistance("hello", "hello"));
        assertEquals("levenshtein one diff", 1, StringUtils.levenshteinDistance("hello", "helo"));
        assertEquals("levenshtein different", 3, StringUtils.levenshteinDistance("kitten", "sitting"));
        assertEquals("levenshtein empty", 5, StringUtils.levenshteinDistance("", "hello"));
        
        // similarityRatio
        assertEquals("similarity identical", 1.0, StringUtils.similarityRatio("hello", "hello"), 0.001);
        assertTrue("similarity similar", StringUtils.similarityRatio("hello", "helo") > 0.7);
        assertTrue("similarity different", StringUtils.similarityRatio("hello", "world") < 0.5);
        
        System.out.println("  Similarity: PASSED\n");
    }
    
    private static void testUtilities() {
        System.out.println("Testing Utilities...");
        
        // slugify
        assertEquals("slugify", "hello-world", StringUtils.slugify("Hello World!"));
        assertEquals("slugify special", "cafe", StringUtils.slugify("Café"));
        
        // randomString
        String random = StringUtils.randomString(10, "abc");
        assertEquals("randomString length", 10, random.length());
        assertTrue("randomString chars", random.matches("[abc]+"));
        
        // randomAlphanumeric
        String randomAlnum = StringUtils.randomAlphanumeric(20);
        assertEquals("randomAlphanumeric length", 20, randomAlnum.length());
        
        // randomHex
        String randomHex = StringUtils.randomHex(16);
        assertEquals("randomHex length", 16, randomHex.length());
        assertTrue("randomHex chars", randomHex.matches("[0-9A-F]+"));
        
        // containsChinese
        assertTrue("containsChinese", StringUtils.containsChinese("你好 World"));
        assertFalse("containsChinese false", StringUtils.containsChinese("Hello World"));
        
        // extractChinese
        assertEquals("extractChinese", "你好", StringUtils.extractChinese("你好 World"));
        
        System.out.println("  Utilities: PASSED\n");
    }
    
    private static void assertFalse(String message, boolean condition) {
        if (!condition) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAILED: " + message);
        }
    }
}

File: Lua/string_utils/string_utils_test.lua
#!/usr/bin/env lua
---
-- String Utilities Test Suite
-- 字符串工具函数测试套件
--
-- 覆盖场景:
-- - 空白处理 (trim, is_empty, is_blank)
-- - 大小写转换 (lower, upper, capitalize, title, case conversion)
-- - 字符串查找 (contains, starts_with, ends_with, find_all)
-- - 字符串替换 (replace_first, replace_last, replace_all, replace_n)
-- - 字符串分割 (split, lines, join)
-- - 字符串填充 (pad_left, pad_right, pad_center)
-- - 字符串截断 (truncate, truncate_words)
-- - 验证函数 (is_alpha, is_digit, is_email, is_url, is_ipv4)
-- - 格式化函数 (format_number, format_bytes, format_time)
-- - 编码解码 (url_encode/decode, html_encode/decode, base64_encode/decode)
-- - 相似度计算 (levenshtein, similarity)
-- - 工具函数 (random, uuid, builder)
--
-- Author: AllToolkit
-- Version: 1.0.0

local path = arg and arg[0] and arg[0]:match("(.*/)") or ""
local mod_path = path .. "mod.lua"

-- 加载模块
local StringUtils = dofile(mod_path)

-- 测试统计
local tests_run = 0
local tests_passed = 0
local tests_failed = 0
local failures = {}

--- 断言函数
local function assert_eq(actual, expected, message)
    tests_run = tests_run + 1
    if actual == expected then
        tests_passed = tests_passed + 1
        return true
    else
        tests_failed = tests_failed + 1
        local msg = message or ("Expected %s, got %s"):format(tostring(expected), tostring(actual))
        table.insert(failures, msg)
        print("  ❌ FAIL: " .. msg)
        return false
    end
end

--- 断言真值
local function assert_true(condition, message)
    tests_run = tests_run + 1
    if condition then
        tests_passed = tests_passed + 1
        return true
    else
        tests_failed = tests_failed + 1
        local msg = message or "Expected true"
        table.insert(failures, msg)
        print("  ❌ FAIL: " .. msg)
        return false
    end
end

--- 断言假值
local function assert_false(condition, message)
    tests_run = tests_run + 1
    if not condition then
        tests_passed = tests_passed + 1
        return true
    else
        tests_failed = tests_failed + 1
        local msg = message or "Expected false"
        table.insert(failures, msg)
        print("  ❌ FAIL: " .. msg)
        return false
    end
end

--- 打印测试组标题
local function test_group(name)
    print("\n📋 " .. name)
    print(string.rep("-", 50))
end

-------------------------------------------------------------------------------
-- 测试：空白处理
-------------------------------------------------------------------------------
test_group("空白处理 (Whitespace Handling)")

assert_eq(StringUtils.trim("  hello  "), "hello", "trim 两端空白")
assert_eq(StringUtils.trim("hello"), "hello", "trim 无空白")
assert_eq(StringUtils.trim("  "), "", "trim 全空白")
assert_true(StringUtils.is_empty(nil), "is_empty(nil)")
assert_true(StringUtils.is_empty(""), "is_empty(\"\")")
assert_false(StringUtils.is_empty(" "), "is_empty(\" \")")
assert_true(StringUtils.is_blank(nil), "is_blank(nil)")
assert_true(StringUtils.is_blank(""), "is_blank(\"\")")
assert_true(StringUtils.is_blank("   "), "is_blank(\"   \")")
assert_false(StringUtils.is_blank(" a "), "is_blank(\" a \")")
assert_eq(StringUtils.trim_left("  hello  "), "hello  ", "trim_left")
assert_eq(StringUtils.trim_right("  hello  "), "  hello", "trim_right")

-------------------------------------------------------------------------------
-- 测试：大小写转换
-------------------------------------------------------------------------------
test_group("大小写转换 (Case Conversion)")

assert_eq(StringUtils.lower("HELLO"), "hello", "lower")
assert_eq(StringUtils.upper("hello"), "HELLO", "upper")
assert_eq(StringUtils.capitalize("hello"), "Hello", "capitalize")
assert_eq(StringUtils.capitalize("HELLO"), "Hello", "capitalize 全大写")
assert_eq(StringUtils.title("hello world"), "Hello World", "title")
assert_eq(StringUtils.to_snake_case("camelCase"), "camel_case", "to_snake_case")
assert_eq(StringUtils.to_snake_case("PascalCase"), "pascal_case", "to_snake_case PascalCase")
assert_eq(StringUtils.to_camel_case("snake_case"), "snakeCase", "to_camel_case")
assert_eq(StringUtils.to_pascal_case("snake_case"), "SnakeCase", "to_pascal_case")

-------------------------------------------------------------------------------
-- 测试：字符串查找
-------------------------------------------------------------------------------
test_group("字符串查找 (String Search)")

assert_true(StringUtils.contains("hello world", "world"), "contains 找到")
assert_false(StringUtils.contains("hello world", "foo"), "contains 未找到")
assert_true(StringUtils.starts_with("hello world", "hello"), "starts_with")
assert_false(StringUtils.starts_with("hello world", "world"), "starts_with false")
assert_true(StringUtils.ends_with("hello world", "world"), "ends_with")
assert_false(StringUtils.ends_with("hello world", "hello"), "ends_with false")

local positions = StringUtils.find_all("abcabcabc", "abc")
assert_eq(#positions, 3, "find_all 找到 3 个")
assert_eq(positions[1], 1, "find_all 位置 1")
assert_eq(positions[2], 4, "find_all 位置 2")
assert_eq(positions[3], 7, "find_all 位置 3")

-------------------------------------------------------------------------------
-- 测试：字符串替换
-------------------------------------------------------------------------------
test_group("字符串替换 (String Replacement)")

assert_eq(StringUtils.replace_first("hello world world", "world", "foo"), 
          "hello foo world", "replace_first")
assert_eq(StringUtils.replace_last("hello world world", "world", "foo"), 
          "hello world foo", "replace_last")
assert_eq(StringUtils.replace_all("hello world world", "world", "foo"), 
          "hello foo foo", "replace_all")
assert_eq(StringUtils.replace_n("a b a b a", "a", "x", 2), 
          "x b x b a", "replace_n")

-------------------------------------------------------------------------------
-- 测试：字符串分割与连接
-------------------------------------------------------------------------------
test_group("分割与连接 (Split & Join)")

local parts = StringUtils.split("a,b,c", ",")
assert_eq(#parts, 3, "split 数量")
assert_eq(parts[1], "a", "split[1]")
assert_eq(parts[2], "b", "split[2]")
assert_eq(parts[3], "c", "split[3]")

local lines = StringUtils.lines("line1\nline2\nline3")
assert_eq(#lines, 3, "lines 数量")

local joined = StringUtils.join({"a", "b", "c"}, "-")
assert_eq(joined, "a-b-c", "join")

-------------------------------------------------------------------------------
-- 测试：字符串填充
-------------------------------------------------------------------------------
test_group("字符串填充 (Padding)")

assert_eq(StringUtils.pad_left("42", 5, "0"), "00042", "pad_left")
assert_eq(StringUtils.pad_right("42", 5, "0"), "42000", "pad_right")
assert_eq(StringUtils.pad_center("42", 7, " "), "  42   ", "pad_center")
assert_eq(StringUtils.pad_center("42", 6, " "), "  42  ", "pad_center 偶数")

-------------------------------------------------------------------------------
-- 测试：字符串截断
-------------------------------------------------------------------------------
test_group("字符串截断 (Truncation)")

assert_eq(StringUtils.truncate("hello world", 8), "hello...", "truncate")
assert_eq(StringUtils.truncate("hi", 10), "hi", "truncate 不需要")
assert_eq(StringUtils.truncate_words("hello world foo bar", 15), "hello world...", "truncate_words")

-------------------------------------------------------------------------------
-- 测试：验证函数
-------------------------------------------------------------------------------
test_group("验证函数 (Validation)")

assert_true(StringUtils.is_alpha("abc"), "is_alpha")
assert_false(StringUtils.is_alpha("abc123"), "is_alpha 含数字")
assert_true(StringUtils.is_digit("123"), "is_digit")
assert_false(StringUtils.is_digit("123a"), "is_digit 含字母")
assert_true(StringUtils.is_alphanumeric("abc123"), "is_alphanumeric")
assert_true(StringUtils.is_email("test@example.com"), "is_email")
assert_false(StringUtils.is_email("invalid"), "is_email 无效")
assert_true(StringUtils.is_url("https://example.com"), "is_url")
assert_true(StringUtils.is_ipv4("192.168.1.1"), "is_ipv4")
assert_false(StringUtils.is_ipv4("256.1.1.1"), "is_ipv4 超限")

-------------------------------------------------------------------------------
-- 测试：格式化函数
-------------------------------------------------------------------------------
test_group("格式化函数 (Formatting)")

assert_eq(StringUtils.format_number(1234567), "1,234,567", "format_number")
assert_eq(StringUtils.format_number(-1234567), "-1,234,567", "format_number 负数")
assert_true(StringUtils.format_bytes(1536):match("^1.50 KB$"), "format_bytes")
assert_eq(StringUtils.format_time(3661), "1:01:01", "format_time")
assert_true(StringUtils.format_time(61):match("^1:01$"), "format_time 无小时")

-------------------------------------------------------------------------------
-- 测试：编码解码
-------------------------------------------------------------------------------
test_group("编码解码 (Encoding/Decoding)")

assert_eq(StringUtils.url_encode("hello world"), "hello+world", "url_encode")
assert_eq(StringUtils.url_decode("hello+world"), "hello world", "url_decode")
assert_eq(StringUtils.html_encode("<script>"), "&lt;script&gt;", "html_encode")
assert_eq(StringUtils.html_decode("&lt;script&gt;"), "<script>", "html_decode")

local encoded = StringUtils.base64_encode("Hello, World!")
local decoded = StringUtils.base64_decode(encoded)
assert_eq(decoded, "Hello, World!", "base64 encode/decode")

-------------------------------------------------------------------------------
-- 测试：相似度
-------------------------------------------------------------------------------
test_group("相似度 (Similarity)")

assert_eq(StringUtils.levenshtein("kitten", "sitting"), 3, "levenshtein")
assert_eq(StringUtils.levenshtein("", "abc"), 3, "levenshtein 空串")
assert_eq(StringUtils.similarity("hello", "hello"), 1.0, "similarity 相同")
assert_true(StringUtils.similarity("hello", "hallo") > 0.5, "similarity 相似")

-------------------------------------------------------------------------------
-- 测试：工具函数
-------------------------------------------------------------------------------
test_group("工具函数 (Utilities)")

local random_str = StringUtils.random(10)
assert_eq(#random_str, 10, "random 长度")

local uuid = StringUtils.uuid()
assert_true(uuid:match("^%x%x%x%x%x%x%x%x%-%x%x%x%x%-%x%x%x%x%-%x%x%x%x%-%x%x%x%x%x%x%x%x%x%x%x%x$"), "uuid 格式")

local builder = StringUtils.builder()
builder:append("Hello"):append(" "):append("World")
assert_eq(builder:to_string(), "Hello World", "builder")

-------------------------------------------------------------------------------
-- 测试：模板与插值
-------------------------------------------------------------------------------
test_group("模板与插值 (Template & Interpolation)")

local template_result = StringUtils.template("Hello, {name}!", { name = "World" })
assert_eq(template_result, "Hello, World!", "template")

local interp_result = StringUtils.interpolate("2 + 2 = ${2 + 2}", {})
assert_eq(interp_result, "2 + 2 = 4", "interpolate 表达式")

-------------------------------------------------------------------------------
-- 测试：边界情况
-------------------------------------------------------------------------------
test_group("边界情况 (Edge Cases)")

assert_true(StringUtils.is_empty(nil), "nil 处理")
assert_eq(StringUtils.trim(nil), nil, "trim nil")
assert_eq(StringUtils.contains(nil, "x"), false, "contains nil")
assert_eq(StringUtils.length(nil), 0, "length nil")
assert_eq(StringUtils.char_at(nil, 1), nil, "char_at nil")
assert_eq(StringUtils.substring(nil, 1, 2), nil, "substring nil")

-------------------------------------------------------------------------------
-- 打印测试结果
-------------------------------------------------------------------------------
print("\n" .. string.rep("=", 50))
print("📊 测试结果")
print(string.rep("=", 50))
print(string.format("总测试数：%d", tests_run))
print(string.format("✅ 通过：%d", tests_passed))
print(string.format("❌ 失败：%d", tests_failed))

if tests_failed > 0 then
    print("\n📝 失败详情:")
    for i, failure in ipairs(failures) do
        print(string.format("  %d. %s", i, failure))
    end
    os.exit(1)
else
    print("\n🎉 所有测试通过!")
    os.exit(0)
end
